### ISSUE TITLE
Set Airflow version in Astronomer UI

### ISSUE THREAD

**Issue by ryw** [astronomer/astronomer#422]
## Create deployment

- [ ] When creating an Airflow deployment, ask what version of Airflow is intended to run.
- [ ] In the background, push the fat image to the docker registry (warm it up for the user)
- [ ] Use this information to block deployments of wrong Airflow versions (https://github.com/astronomer/astro-cli/issues/269)

## Update deployment

- [ ] If user changes Airflow version, we push fat image for that Airflow version to their Docker (but doesn't do anything to currently running version) - the deployment should have a banner "Deployed version doesn't match configured version" until new deployment is pushed

**Comment by schnie** [comment]
Recapping convo with @ryw:

For this first version, this feature should probably have more of a "guardrail" feel to it. We can't yet truly block a deployment if they are not pushing a "certified" image that we can grab the version from. This will most likely require more infrastructure than we really want to get into now.

We could allow a user to _optionally_ set an airflow version on the deployment and we could use that value to run a crude check against the `FROM` line of the `Dockerfile` before deploying. This doesn't work in all situations and would pave over an important escape hatch for us, so we should add a flag to disable this check if someone is using a custom image.

**Comment by ryw** [comment]
We could eventually add another box where they can define what the FROM line should be in the dockerfile, and we could pre-warm even with their custom image.

**Comment by schnie** [comment]
Capturing some convos from slack and other thoughts here - 

**First thing - Actually having multiple supported versions of Airflow.**
Maybe there is a way to pull the bulk of the base dependencies into an `airflow-base` image, then build our existing airflow base images off of that one. Then we could start supporting multiple LTS versions of airflow that make any dependency adjustments and install the appropriate version of airflow on top (and then our standard `-onbuild`s on top of that). This feels like a way to support multiple versions of airflow that are kept up to date with all our add-ons (security manager, theme, etc). @ashb @andriisoldatenko @dimberman any thoughts on this? Not sure how this plays with us using an official airflow image. Wondering if our needs for our product will exactly line up with community image or not.

**Second thing - Pre-warming Astronomer Deployment Docker Repositories**
@ashb brought up the idea of using [Cross Repository Blob Mounting](https://docs.docker.com/registry/spec/api/) to pre-mount docker layers into a repository, before a user attempts a code push. This feels like a cleaner alternative to `docker push`ing an image every time we create a new deployment, but will require us to pre-populate a central repository with our airflow images. Maybe this is a helm `post-install` hook that runs a kubernetes `Job`, that pulls the airflow images (all versions) from dockerhub (or private registry https://github.com/astronomer/helm.astronomer.io/blob/master/values.yaml#L51-L53) and pushes them to the fresh internal astronomer repository (so their layers can be mounted later).

With the base blobs in place, we can have Houston make the Cross Repository Blob Mount API call for every `createDeployment` mutation that comes through.

Then when an end-user goes to do their first `astro deploy`, the only thing that should need to be actually uploaded (on `docker push`) will be the layers that contain any of their custom dependencies and DAG code.

Cross Repository Blob Mount - https://docs.docker.com/registry/spec/api/
Slack convos - https://astronomerteam.slack.com/archives/C02NZELT2/p1566508479256900, https://astronomerteam.slack.com/archives/C02NZELT2/p1566509276264300
Cross-linking https://github.com/astronomer/astronomer/issues/264 since its related.

**Comment by ashb** [comment]
On the subject of pre-warming, I tested something. Running a local docker registry I tried pushing these two images:

```bash 
$ docker tag explosive-apogee-6556/airflow localhost:5000/explosive-apogee-6556/airflow:cli-1
$ time docker push localhost:5000/explosive-apogee-6556/airflow:cli-1
The push refers to repository [localhost:5000/explosive-apogee-6556/airflow]
c4b79e0a97ed: Pushed
ccb1b0eb26cf: Pushed
85e5455d4943: Pushed
6cc3ceb52fb6: Pushed
2ff36a10ff34: Pushed
172adc29457f: Pushed
7af4d3d902bd: Pushed
88cd5f2c0567: Pushed
499ab822df93: Pushed
93ac506b97dc: Pushed
c4f690806365: Pushed
1bfeebd65323: Pushed
cli-1: digest: sha256:bf50137c59f858d9646bd67f051fdf888ed09b10df851aa372abbb3e0a2db33a size: 2824
docker push localhost:5000/explosive-apogee-6556/airflow:cli-1  0.16s user 0.10s system 0% cpu 1:24.87 total
```

Then tag it as another image and push up

```
$ docker tag explosive-apogee-6556/airflow localhost:5000/other/airflow:cli-1
$ time docker push localhost:5000/other/airflow:cli-1
The push refers to repository [localhost:5000/other/airflow]
c4b79e0a97ed: Mounted from explosive-apogee-6556/airflow
ccb1b0eb26cf: Mounted from explosive-apogee-6556/airflow
85e5455d4943: Mounted from explosive-apogee-6556/airflow
6cc3ceb52fb6: Mounted from explosive-apogee-6556/airflow
2ff36a10ff34: Mounted from explosive-apogee-6556/airflow
172adc29457f: Mounted from explosive-apogee-6556/airflow
7af4d3d902bd: Mounted from explosive-apogee-6556/airflow
88cd5f2c0567: Mounted from explosive-apogee-6556/airflow
499ab822df93: Mounted from explosive-apogee-6556/airflow
93ac506b97dc: Mounted from explosive-apogee-6556/airflow
c4f690806365: Mounted from explosive-apogee-6556/airflow
1bfeebd65323: Mounted from explosive-apogee-6556/airflow
cli-1: digest: sha256:bf50137c59f858d9646bd67f051fdf888ed09b10df851aa372abbb3e0a2db33a size: 2824
docker push localhost:5000/other/airflow:cli-1  0.06s user 0.06s system 11% cpu 1.048 total
```

And then this form the [docs](https://docs.docker.com/registry/spec/api/#cross-repository-blob-mount) is I think relevant:

> A blob may be mounted from another repository that the client has read access to, removing the need to upload a blob already known to the registry

So the reason we don't get any mounting/sharing going on our registry is probably because the JWT token we provide to a registry is limited to a single repo at a time. If we instead made the token include permission (at least pull) to all the repos the user had, plus a new "base-image" repo, then the we won't have to do a manual cross-repo blob mount ourselves

**Comment by schnie** [comment]
@ashb good catch. That actually sounds like a relatively small change on the Houston side of things then. We'd really just have to append an item or two to this list right? https://github.com/astronomer/houston-api/blob/master/src/resolvers/types/deployment/index.js#L94-L98

**Comment by ashb** [comment]
Yup, think so.
